examples/cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,257 @@
+            raw_orders[4*this->n_orders + 0] = it->alphas;
+            raw_orders[4*this->n_orders + 1] = it->alpha;
+            raw_orders[4*this->n_orders + 2] = it->logxir;
+            raw_orders[4*this->n_orders + 3] = it->logxif;
+        }
+        //this->n_bins = bins.size();
+        // cast bins
+        double *raw_bins = new double[bins.size()];
+        for (auto it = bins.cbegin(); it != bins.cend(); ++it, ++this->n_bins)
+            raw_bins[this->n_bins] = *it;
+        // now, we got all we need
+        this->raw = pineappl_grid_new(lumi.raw, this->n_orders, raw_orders, this->n_bins - 1, raw_bins, key_val.raw);
+        // clean up
+        delete[] raw_orders;
+        delete[] raw_bins;
+    }
+
+    /** @brief destructor */
+    ~Grid(){ pineappl_grid_delete(this->raw); }
+
+    /**
+     * @brief Fill grid for the given parameters
+     * @param x1 first momentum fraction
+     * @param x2 second momentum fraction
+     * @param q2 scale
+     * @param order order index
+     * @param observable observable value
+     * @param lumi luminosity index
+     * @param weight weight
+     */
+    void fill(double x1, double x2, double q2, size_t order, double observable, size_t lumi, double weight) const {
+        pineappl_grid_fill(this->raw, x1, x2, q2, order, observable, lumi, weight);
+    }
+
+    /**
+     * @brief perform a convolution of the grid with PDFs
+     * @param pdg_id particle ID
+     * @param pdf PDF
+     * @param xi_ren renormalization scale variation
+     * @param xi_fac factorization scale variation
+     * @return prediction for each bin
+     * @return prediction for each bin
+     */
+    std::vector<double> convolute_with_one(int32_t pdg_id, LHAPDF::PDF *pdf, double xi_ren = 1.0, double xi_fac = 1.0 ) const {
+        std::vector<bool> order_mask(this->n_orders, true);
+        std::vector<bool> lumi_mask(this->n_lumis, true);
+        return this->convolute_with_one(pdg_id, pdf, order_mask, lumi_mask, xi_ren, xi_fac);
+    }
+
+    /**
+     * @brief perform a convolution of the grid with PDFs
+     * @param pdg_id particle ID
+     * @param pdf PDF
+     * @param order_mask order mask
+     * @param lumi_mask luminosity mask
+     * @param xi_ren renormalization scale variation
+     * @param xi_fac factorization scale variation
+     * @return prediction for each bin
+     */
+    std::vector<double> convolute_with_one(int32_t pdg_id, LHAPDF::PDF *pdf, const std::vector<bool>& order_mask, const std::vector<bool>& lumi_mask, double xi_ren = 1.0, double xi_fac = 1.0) const {
+        // prepare LHAPDF stuff
+        auto xfx = [](int32_t id, double x, double q2, void* pdf) {
+            return static_cast <LHAPDF::PDF*> (pdf)->xfxQ2(id, x, q2);
+        };
+        auto alphas = [](double q2, void* pdf) {
+            return static_cast <LHAPDF::PDF*> (pdf)->alphasQ2(q2);
+        };
+        // cast order_mask
+        bool *raw_order_mask = new bool[this->n_orders];
```

Comment:
Delete these lines and pass `order_mask.data()` and `lumi_mask.data()` instead to the CAPI.

Reply:
I can't - because `PineAPPL.hpp:217:41: error: use of deleted function ‘void std::vector<bool, _Alloc>::data() [with _Alloc = std::allocator<bool>]’
  217 |         auto *raw_mask = order_mask.data();` seems like `vector<bool>` is special and even considered evil ...